### PR TITLE
Backend

### LINKED ISSUE #66: Single Neuron
- [x] 1. Neuron class
- [x] 2. Weigthed sum (using Math.Numerics)
- [x] 3. Activation function: tanh, sigmoid, ReLu

### LINKED ISSUE #68: Unit Test for Single Neuron
- [x] Happy path
- [x] All weights zero -> weightSum zero
- [x] Positive, negative weights
- [x] Activation functions (tanh, sigmoid, ReLu)

### PR DESCRIPTION
**Pull request by molbazsa**
I think we should merge the Linear Regression code to main so that everything is up to date.

### CODE REVIEW

**Comment by molbazsa** [comment]
Neuron created

**Comment by molbazsa** [comment]
Activation functions created:
- Sigmoid
- Tanh
- ReLU

File: LinearRegression/LinearRegressionBackend/DataProvider/Numerical.cs [hunk 1]
```diff
@@ -16,104 +22,223 @@ public static class Numerical
...
+            if (axis == 0)
+            {
+                for (int i = 0; i < matrix.GetLength(0); i++)
+                {
+                    matrix_array[i] = matrix[i][axis];
+                }
+                SortArray(matrix_array);
+                if (matrix_array.Length % 2 == 0)
+                {
+                    double egy = matrix_array[(int)middle - 1];
```
Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/DataProvider/Numerical.cs:123]: Use english variable name

Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/DataProvider/Numerical.cs]: Use english variable name

File: LinearRegression/LinearRegressionBackend/DataProvider/Numerical.cs [hunk 2]
```diff
@@ -16,104 +22,223 @@ public static class Numerical
...
+        /// <param name="matrix">The matrix, containing the values.</param>
+        /// <param name="axis">The axis whose median we want to calculate. 0 for X, 1 for Y.</param>
+        /// <returns>Returns the Median of all values, from the given axis.</returns>
         public static double Median(double[][] matrix, int axis)
         {
-            double result = 0;
-
-            return result;
+            double[] matrix_array = new double[matrix.GetLength(0)];
+            double middle = matrix_array.GetLength(0) / 2;
```
Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/DataProvider/Numerical.cs]: Change `middle` type from double to int

File: LinearRegression/LinearRegressionBackend/DataProvider/Numerical.cs [hunk 3]
```diff
@@ -16,104 +22,223 @@ public static class Numerical
...
+                    double egy = matrix_array[(int)middle - 1];
+                    double ketto = matrix_array[(int)middle];
+                    return ((matrix_array[(int)middle - 1] + matrix_array[(int)middle]) / 2);
+                }
+                else
+                {
+                    return (matrix_array[(int)middle]);
+                }
+            }
+            else
```
Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/DataProvider/Numerical.cs:121]: Remove code duplication

File: LinearRegression/LinearRegressionBackend/DataProvider/Numerical.cs [hunk 4]
```diff
@@ -16,104 +22,223 @@ public static class Numerical
...
+        /// <param name="axis">The axis whose standard deviation we want to calculate. 0 for X, 1 for Y.</param>
+        /// <returns>Returns the Standard deviation of all values, from the given axis.</returns>
         public static double StandardDeviation(double[][] matrix, int axis)
         {
-            double result = 0;
-
-            return result;
+            double mean = Numerical.Mean(matrix,axis);
+            double result=0;
+            //double[] matrix_array = new double[matrix.GetLength(0)];
```
Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/DataProvider/Numerical.cs]: Remove zombie code

File: LinearRegression/LinearRegressionBackend/DataProvider/Numerical.cs [hunk 5]
```diff
@@ -16,104 +22,223 @@ public static class Numerical
         /// <returns>Returns with an array.</returns>
         public static double[] Arange(double start, double stop, double step)
         {
-            double[] array = null;
-
+            decimal stepD = (decimal)step;
```
Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/DataProvider/Numerical.cs:25]: Change decimal type to double

File: LinearRegression/LinearRegressionBackend/MLModel/IMLModel.cs [hunk 6]
```diff
@@ -7,14 +7,14 @@ namespace LinearRegressionBackend.MLModel
...
-        /// <summary>
+       // <summary>
         /// Trains the model for a fixed number of epochs.
         /// </summary>
         /// <param name="inputData">The matrix representation of the input data.</param>
         /// <param name="targetData">The array representation of the target data.</param>
         /// <param name="epochs">The number of iterations.</param>
         /// <returns>Returns with the training history.</returns>
-        History Fit(double[][] inputData, double[] targetData, int epochs=1);
+       // History Fit(double[][] inputData, double[] targetData, int epochs=1);
```
Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/MLModel/IMLModel.cs]: Remove zombie code. Place summary to `Train` method.

File: LinearRegression/LinearRegressionBackend/MLNeuralNetwork/Neuron.cs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+using MathNet.Numerics.LinearAlgebra;
+
+namespace LinearRegressionBackend.MLNeuralNetwork {
+    public class Neuron {
+        private Vector<double> _Weights;
+        private double _Bias;
+        private IActivationFunction _ActivationFunction;
+
+        public Neuron(Vector<double> weights, IActivationFunction activationFunction) {
```
Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/MLNeuralNetwork/Neuron.cs]: Initialize bias as well

File: LinearRegression/LinearRegressionBackend/Program.cs [hunk 8]
```diff
@@ -35,7 +35,47 @@ static void Main(string[] args)
...
+            myArray4[1][0] = 3;myArray4[1][1] = 93;
+            myArray4[2][0] = 5;myArray4[2][1] = 98;
+            myArray4[3][0] = 7;myArray4[3][1] = 89;
+            myArray4[4][0] = 9;myArray4[4][1] = 88;
+
+            double[,] multiDimensionalArray = { { 1, 2 }, { 3, 4 }, { 3, 4 }, { 5, 7 }, { 9, 10 } };
+
+            double testDongle =Numerical.StandardDeviation(myArray4,1);
+            Console.WriteLine(testDongle);
+            /*
```
Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/Program.cs:62]: Remove zombie code

File: LinearRegression/Numrical_uTest/NumericalTest.cs [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+            }
+            
+        }
+
+        [Test]
+        public void Mean_DoubleTwoDimensionalMatrix_ReturnsSuccessfully()
+        {
+            // Arrange
+            double[][] testMatrix = new double[5][];
+            testMatrix[0] = new double[2];
```
Comment (SzilvasiPeter) [inline LinearRegression/Numrical_uTest/NumericalTest.cs:39]: Simplify `testMatrix` initialization.

**Review by SzilvasiPeter — COMMENTED** [review]
Review findings were added

**Review by SzilvasiPeter — APPROVED** [review]

**Pull request merged**